File: src/main/java/Pandigital.java
import fj.*;
import fj.data.List;
import fj.data.Stream;
import fj.function.Integers;

import static fj.data.Stream.*;
import static fj.Show.*;

public class Pandigital {
    public static void main(String[] args) {
        List<Integer> list = count().toList().nub();
        intShow.println(Integers.sum(list));
        listShow(intShow).println(list);
    }

    static Stream<Stream<Character>> permute(Stream<Character> stream) {
        return stream.bind(x -> {
            Stream<Character> s = stream.filter(a -> !a.equals(x));
            if(s.isEmpty())
                return single(single(x));
            else
                return permute(s).map(xs -> cons(x, () -> xs));
        });
    }

    static Stream<Integer> count() {
        Stream<Stream<Character>> streams = permute(fromString("123456789"));
        return streams.map(Pandigital::digitToInt)
                .map(i -> quotRem(i, 100000))
                .filter(p2 -> check(10, p2) || check(100, p2))
                .map(P2::_1);
    }

    static int digitToInt(Stream<Character> stream) {
        return stream.foldLeft((s, n) -> s * 10 + Character.getNumericValue(n), 0);
    }

    static P2<Integer, Integer> quotRem(Integer x, Integer y) {
        return P.lazy(() -> x / y, () -> x % y);
    }

    static boolean check(int n, P2<Integer, Integer> p2) {
        P2<Integer, Integer> p = quotRem(p2._2(), n);
        return p2._1() == p._1() * p._2();
    }
}

File: src/main/kotlin/Pandigital.kt
import fj.F
import fj.P
import fj.P2
import fj.Show.intShow
import fj.Show.listShow
import fj.data.Stream
import fj.function.Integers

/**
 * Created by yume on 16-12-12.
 */

fun main(args: Array<String>) {
//    println(int_seq(listOf(1, 2, 3, 4)))
    val list = count().toList().nub()
    intShow.println(Integers.sum(list))
    listShow<Int>(intShow).println(list)
}

internal fun permute(stream: Stream<Char>): Stream<Stream<Char>> =
        stream.bind { x ->
            val s = stream.filter { a -> a != x }
            if (s.isEmpty)
                Stream.single(Stream.single(x))
            else
                permute(s).map { xs -> Stream.cons<Char>(x) { xs } }
        }

internal fun count(): Stream<Int> =
        permute(Stream.fromString("123456789"))
                .map { digitToInt(it) }
                .map { i -> quotRem(i, 100000) }
                .filter { p2 -> check(10, p2) || check(100, p2) }
                .map { it._1() }

internal fun digitToInt(stream: Stream<Char>): Int =
        stream.foldLeft({ s, n -> s * 10 + Character.getNumericValue(n) }, 0)

internal fun quotRem(x: Int, y: Int): P2<Int, Int> = P.p(x / y, x % y)

internal fun check(n: Int, p2: P2<Int, Int>): Boolean {
    val p = quotRem(p2._2(), n)
    return p2._1() == p._1() * p._2()
}